### PR TITLE
[velux] Add support for vane/slat position

### LINKED ISSUE #12608: [velux] Add support for slat position of blinds
As stated on the forum (see link below) a user is requesting to add support for monitoring and controlling the slat position on (Somfy) Venetian blinds. i.e. Adding support for Functional Parameter 1 (FP1)

https://community.openhab.org/t/velux-new-openhab2-binding-feedback-welcome/32926/1350

### PR DESCRIPTION
**Pull request by andrewfg**
Fixes #12608

On the community forum, a user has requested to add support for reading and setting the vane/slat position of Venetian type blinds. See this https://community.openhab.org/t/velux-new-openhab2-binding-feedback-welcome/32926/1350

This PR does that by means of the following..
- Added a new Channel `vanePosition` of type Dimmer to the Thing type RollerShutter.
- Added support for reading and setting Functional Parameters FP1 .. FP4 and mapping to/from `vanePosition` Channel.
- Cleaned up and extended the trace logging for Functional Parameters FP1 .. FP4 values.
- Fixed compiler PMD.CompareObjectsWithEquals warnings.
- Fixed a minor compiler NPE warning.
- Added Junit test suite for new and existing functionality.
- Improved logging to enable easier tracking of state sequences via RegEx text filters.
- Many code optimisations based on learnings from the JUnit tests and the improved logging.

Signed-off-by: Andrew Fiddian-Green <software@whitebear.ch>

### CODE REVIEW

**Comment by andrewfg** [comment]
The JAR file (for testing) is here..

[org.openhab.binding.velux-3.3.0-SNAPSHOT.jar.zip](https://github.com/openhab/openhab-addons/files/8505669/org.openhab.binding.velux-3.3.0-SNAPSHOT.jar.zip)

**Comment by andrewfg** [comment]
Latest JARS are here..

1. New API version: [org.openhab.binding.velux-3.3.0-SNAPSHOT.jar.zip](https://github.com/openhab/openhab-addons/files/8695644/org.openhab.binding.velux-3.3.0-SNAPSHOT.jar.zip)

2. Legacy API version: [org.openhab.binding.velux-3.3.0-SNAPSHOT.jar.zip](https://github.com/openhab/openhab-addons/files/8695651/org.openhab.binding.velux-3.3.0-SNAPSHOT.jar.zip)

**Comment by jlaur** [comment]
> the CI Build on this PR now fails due to an apparent inconsistency between the new spotless formatter output (see #13027) and the (old) code format analyser.

It seems to have completed without failure now.

**Comment by andrewfg** [comment]
> seems to have completed without failure now.

Yes. It was probably a synchronisation or caching issue.

File: bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/discovery/VeluxDiscoveryService.java [hunk 1]
```diff
@@ -68,9 +68,8 @@ public class VeluxDiscoveryService extends AbstractDiscoveryService implements R
 
     // Private
 
-    @SuppressWarnings("PMD.CompareObjectsWithEquals")
     private void updateLocalization() {
-        if (localization == Localization.UNKNOWN && localeProvider != null && i18nProvider != null) {
+        if (localization.equals(Localization.UNKNOWN) && (localeProvider != null) && (i18nProvider != null)) {
```
Comment (jlaur) [inline bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/discovery/VeluxDiscoveryService.java:72]:
```suggestion
        if (Localization.UNKNOWN.equals(localization) && (localeProvider != null) && (i18nProvider != null)) {
```

File: bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/factory/VeluxHandlerFactory.java [hunk 2]
```diff
@@ -125,9 +125,8 @@ private void updateBindingState() {
         });
     }
 
-    @SuppressWarnings("PMD.CompareObjectsWithEquals")
     private void updateLocalization() {
-        if (localization == Localization.UNKNOWN && localeProvider != null && i18nProvider != null) {
+        if (localization.equals(Localization.UNKNOWN) && (localeProvider != null) && (i18nProvider != null)) {
```
Comment (jlaur) [inline bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/factory/VeluxHandlerFactory.java:129]:
```suggestion
        if (Localization.UNKNOWN.equals(localization) && (localeProvider != null) && (i18nProvider != null)) {
```

File: bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/things/VeluxKLFAPI.java [hunk 3]
```diff
@@ -290,11 +291,11 @@ public enum Command {
...
         GW_GET_ACTIVATION_LOG_LINE_REQ((short) 0x0504, "Request line from activation log."),
-        GW_GET_ACTIVATION_LOG_LINE_CFM((short) 0x0505, "Confirm line from activation log."),
-        GW_ACTIVATION_LOG_UPDATED_NTF((short) 0x0506, "Confirm line from activation log."),
+        GW_GET_ACTIVATION_LOG_LINE_CFM((short) 0x0505, "Acknowldge to confirm line from activation log."),
+        GW_ACTIVATION_LOG_UPDATED_NTF((short) 0x0506, "Notification to confirm line from activation log."),
         GW_GET_MULTIPLE_ACTIVATION_LOG_LINES_REQ((short) 0x0507, "Request lines from activation log."),
-        GW_GET_MULTIPLE_ACTIVATION_LOG_LINES_NTF((short) 0x0508, "Error log  data from activation log."),
-        GW_GET_MULTIPLE_ACTIVATION_LOG_LINES_CFM((short) 0x0509, "Confirm lines from activation log."),
+        GW_GET_MULTIPLE_ACTIVATION_LOG_LINES_NTF((short) 0x0508, "Notification error log  data from activation log."),
+        GW_GET_MULTIPLE_ACTIVATION_LOG_LINES_CFM((short) 0x0509, "Acknowldge to confirm lines from activation log."),
```
Comment (jlaur) [inline bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/things/VeluxKLFAPI.java:298]:
```suggestion
        GW_GET_MULTIPLE_ACTIVATION_LOG_LINES_CFM((short) 0x0509, "Acknowledge to confirm lines from activation log."),
```

**Review by jlaur — CHANGES_REQUESTED** [review]
Thanks for adding this feature. I have provided some feedback in form of one small comment per every second file. :-)

File: bundles/org.openhab.binding.velux/README.md [hunk 4]
```diff
@@ -38,6 +38,10 @@ The binding will automatically discover Velux Bridges within the local network,
 Once a Velux Bridge has been discovered, you will need to enter the `password` Configuration Parameter (see below) before the binding can communicate with it.
 And once the Velux Bridge is fully configured, the binding will automatically discover all its respective scenes and actuators (like windows and rollershutters), and place them in the Inbox.
 
+Note: When the KLF200 hub is started it provides a temporary private WiFi Access Point for initial configuration.
```
[the remark this answers is not in the document]
Comment (andrewfg) [reply bundles/org.openhab.binding.velux/README.md:41]: Ok.

File: bundles/org.openhab.binding.velux/README.md [hunk 5]
```diff
@@ -154,6 +158,23 @@ The `position` Channel indicates the open/close state of the window (resp. rolle
...
+|--------------|---------------|-------------------------------------------------|
+| position     | Rollershutter | Actual position of the window or device.        |
+| limitMinimum | Rollershutter | Minimum limit position of the window or device. |
+| limitMaximum | Rollershutter | Maximum limit position of the window or device. |
+| vanePosition | Dimmer        | Vane position of a Venetian blind.              |
+
+The `position`, `limitMinimum`, and `limitMaximum` are the same as described above for "window" Things.
+
+The `vanePosition` Channel only applies to Venetian blinds that have tiltable slats.
+It can only have a valid position value, if the main `position` of the Thing is fully down.
```
[the remark this answers is not in the document]
Comment (andrewfg) [reply bundles/org.openhab.binding.velux/README.md:175]: Ok.

File: bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/discovery/VeluxDiscoveryService.java (hunk 1 above)
Comment (andrewfg) [reply bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/discovery/VeluxDiscoveryService.java:72]: Ok.

File: bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/factory/VeluxHandlerFactory.java (hunk 2 above)
Comment (andrewfg) [reply bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/factory/VeluxHandlerFactory.java:129]: Ok.

File: bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/things/VeluxKLFAPI.java [hunk 6]
```diff
@@ -290,11 +291,11 @@ public enum Command {
         GW_CLEAR_ACTIVATION_LOG_REQ((short) 0x0502, "Request clear all data  in activation log."),
         GW_CLEAR_ACTIVATION_LOG_CFM((short) 0x0503, "Confirm clear all data  in activation log."),
         GW_GET_ACTIVATION_LOG_LINE_REQ((short) 0x0504, "Request line from activation log."),
-        GW_GET_ACTIVATION_LOG_LINE_CFM((short) 0x0505, "Confirm line from activation log."),
-        GW_ACTIVATION_LOG_UPDATED_NTF((short) 0x0506, "Confirm line from activation log."),
+        GW_GET_ACTIVATION_LOG_LINE_CFM((short) 0x0505, "Acknowldge to confirm line from activation log."),
```
[the remark this answers is not in the document]
Comment (andrewfg) [reply bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/things/VeluxKLFAPI.java:294]: Ok.

File: bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/things/VeluxKLFAPI.java (hunk 3 above)
Comment (andrewfg) [reply bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/things/VeluxKLFAPI.java:298]: Ok.

File: bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/things/VeluxProduct.java [hunk 7]
```diff
@@ -216,6 +228,12 @@ public String toString() {
 
     // Class helper methods
 
+    /**
+     * Return the product unique index.
+     * Either the serial number (for normal Velux devices), or its name (for e.g. Somfy devices).
+     *
+     * @return ..
```
[the remark this answers is not in the document]
Comment (andrewfg) [reply bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/things/VeluxProduct.java:235]: Ok.

File: bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/handler/ChannelVShutterPosition.java [hunk 8]
```diff
@@ -148,7 +147,7 @@ private ChannelVShutterPosition() {
         LOGGER.trace("handleCommand(): scene name is {}.", sceneName);
         VeluxScene thisScene2 = thisBridgeHandler.bridgeParameters.scenes.getChannel().existingScenes
                 .get(new SceneName(sceneName));
-        if (thisScene2 == VeluxScene.UNKNOWN) {
+        if (thisScene2.equals(VeluxScene.UNKNOWN)) {
```
[the remark this answers is not in the document]
Comment (andrewfg) [reply bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/handler/ChannelVShutterPosition.java:150]: Ok.

File: bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/things/VeluxProduct.java [hunk 9]
```diff
@@ -385,4 +403,103 @@ public int getDisplayPosition() {
...
+    /**
+     * Set the Functional Parameters. Calls setProductAllowedValues() so that any values that are not allowed by normal
+     * products will be replaced by the 'undefined' (i.e. 0xF7FF) value.
+     *
+     * @param newFunctionalParameters the new values of the Functional Parameters.
+     * @return <b>modified</b> if any of the Functional Parameters have been changed.
+     */
+    public boolean setFunctionalParameters(@Nullable FunctionalParameters newFunctionalParameters) {
+        if (newFunctionalParameters == null) {
+            return false;
```
[the remark this answers is not in the document]
Comment (andrewfg) [reply bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/things/VeluxProduct.java:426]: No. Passing a null value of `newFunctionalParameters` is a special case. If there is no new value of functional parameters it means that nothing has been provided to override any prior value.

However I will update the Java Doc..

File: bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/handler/ChannelActuatorPosition.java [hunk 10]
```diff
@@ -138,48 +171,60 @@ private ChannelActuatorPosition() {
...
+                    }
+                    break;
+                case CHANNEL_ACTUATOR_STATE:
+                    if (command instanceof OnOffType) {
+                        mainPosition = OnOffType.OFF.equals(command) ^ veluxActuator.isInverted()
+                                ? new VeluxProductPosition(PercentType.ZERO)
+                                : new VeluxProductPosition(PercentType.HUNDRED);
+                    }
+                    break;
+                default:
```
[the remark this answers is not in the document]
Comment (andrewfg) [reply bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/handler/ChannelActuatorPosition.java:216]: Ok.

File: bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/handler/utils/Thing2VeluxActuator.java [hunk 11]
```diff
@@ -135,12 +135,11 @@ public Thing2VeluxActuator(VeluxBridgeHandler thisBridgeHandler, ChannelUID this
      *
      * @return <b>bridgeProductIndex</B> for accessing the Velux device (or ProductBridgeIndex.UNKNOWN if not found).
      */
-    @SuppressWarnings("PMD.CompareObjectsWithEquals")
     public ProductBridgeIndex getProductBridgeIndex() {
-        if (thisProduct == VeluxProduct.UNKNOWN) {
+        if (thisProduct.equals(VeluxProduct.UNKNOWN)) {
```
[the remark this answers is not in the document]
Comment (andrewfg) [reply bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/handler/utils/Thing2VeluxActuator.java:139]: Ok.

File: bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/handler/utils/Thing2VeluxActuator.java [hunk 12]
```diff
@@ -135,12 +135,11 @@ public Thing2VeluxActuator(VeluxBridgeHandler thisBridgeHandler, ChannelUID this
...
             mapThing2Velux();
         }
-        if (thisProduct == VeluxProduct.UNKNOWN) {
+        if (thisProduct.equals(VeluxProduct.UNKNOWN)) {
```
[the remark this answers is not in the document]
Comment (andrewfg) [reply bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/handler/utils/Thing2VeluxActuator.java:142]: Ok.

File: bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/handler/utils/Thing2VeluxActuator.java [hunk 13]
```diff
@@ -152,9 +151,8 @@ public ProductBridgeIndex getProductBridgeIndex() {
      *
      * @return <b>isKnown</B> as boolean.
      */
-    @SuppressWarnings("PMD.CompareObjectsWithEquals")
     public boolean isKnown() {
-        return (!(this.getProductBridgeIndex() == ProductBridgeIndex.UNKNOWN));
+        return (!(this.getProductBridgeIndex().equals(ProductBridgeIndex.UNKNOWN)));
```
[the remark this answers is not in the document]
Comment (andrewfg) [reply bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/handler/utils/Thing2VeluxActuator.java:155]: Ok.

File: bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/handler/utils/Thing2VeluxActuator.java [hunk 14]
```diff
@@ -164,12 +162,11 @@ public boolean isKnown() {
      *
      * @return <b>isInverted</B> for handling of values of the Velux device (or false if not found)..
      */
-    @SuppressWarnings("PMD.CompareObjectsWithEquals")
     public boolean isInverted() {
-        if (thisProduct == VeluxProduct.UNKNOWN) {
+        if (thisProduct.equals(VeluxProduct.UNKNOWN)) {
```
[the remark this answers is not in the document]
Comment (andrewfg) [reply bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/handler/utils/Thing2VeluxActuator.java:166]: Ok.

File: bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/handler/utils/Thing2VeluxActuator.java [hunk 15]
```diff
@@ -164,12 +162,11 @@ public boolean isKnown() {
...
             mapThing2Velux();
         }
-        if (thisProduct == VeluxProduct.UNKNOWN) {
+        if (thisProduct.equals(VeluxProduct.UNKNOWN)) {
```
[the remark this answers is not in the document]
Comment (andrewfg) [reply bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/handler/utils/Thing2VeluxActuator.java:169]: Ok.

**Comment by andrewfg** [comment]
@jlaur just a status update for you..
- The user who had requested these changes has not yet finally tested and confirmed that everything is working.
- However in the meantime, I took some time to write a JUnit suite for the functions I have added.
- And based on those tests, I discovered some improvement potential, and changed the code accordingly.
- Note: the existing code is not covered by JUnit tests; that is a task for the future :)

I will send you another update when the user has confirmed the functionality.

**Comment by andrewfg** [comment]
The latest JAR file is here..

[org.openhab.binding.velux-3.4.0-SNAPSHOT.jar.zip](https://github.com/openhab/openhab-addons/files/9048221/org.openhab.binding.velux-3.4.0-SNAPSHOT.jar.zip)

**Comment by andrewfg** [comment]
@jlaur a further update..

The user who requested this new feature has Somfy actuators rather than Velux ones. These are notorious for being rather quirky. From the user's logs I found out that his Somfy actuators are producing undocumented State values in response to the ‘old’ API calls. Specifically, a standard Velux actuator returns State:5 when it is not moving, whereas his Somfy actuators return State:45 …

This issue is actually not problematic for the newly introduced vane position channel because I am now anyway using the ‘new’ API calls to read those values, and this API returns the State by a different means. However it is (and always was) a problem for the main position channel (on Somfy actuators).

So the PR must now also become a bug fix for the wrong State on the main position specifically onSomfy actuators, in addition to being an enhancement adding the vane position.

Bear with me...

**Comment by andrewfg** [comment]
The latest JAR is here..

[org.openhab.binding.velux-3.4.0-SNAPSHOT.jar.zip](https://github.com/openhab/openhab-addons/files/9079300/org.openhab.binding.velux-3.4.0-SNAPSHOT.jar.zip)

**Comment by andrewfg** [comment]
Latest JAR is here..

[org.openhab.binding.velux-3.4.0-SNAPSHOT.jar.zip](https://github.com/openhab/openhab-addons/files/9103096/org.openhab.binding.velux-3.4.0-SNAPSHOT.jar.zip)

**Comment by jlaur** [comment]
@andrewfg, congrats with reaching 100 commits in this PR. ;)

**Comment by andrewfg** [comment]
> congrats with reaching 100 commits in this PR

@jlaur you probably are wondering why there are so many commits, so following is an explanation of what I have been doing..

The PR was initiated by a user request from someone who has a type of shade which I myself do not physically have. So the testing cycle was quite challenging, where I would make a new JAR for the user to install, test, and send me the logs. Such a testing cycle can take many days, if not months (he was on a long vacation), and needed trials of various different hypotheses before one can hit on the right solution.

So to speed things up, I decided instead to create a JUnit test suite where I can inject data captured from the user's logs, and I can try and test very quickly many solution hypotheses, without having to go back to the user each time. This is way faster.

Furthermore, the existing code base has a 'humungous' amount of logging output, which makes it very difficult to read and discern the real functionality sequences. So I made quite a few changes in the logging in order to help me. I tried not to remove the logging of the original author since he had his own reasons for this, but I changed and added some logging so I could search the log files better using a RegEx text filter.

However...

I can now run the JUnit test suite on data from my own devices. And I could see areas where the functionality is not optimal. And I could test edge cases and potential state combinations and transitions which had not previously been checked properly. And with the better logging and RegEx filtering I can see cases where the existing code does not handle the state transitions as one would expect. And. And.

So this has resulted in me making many code improvements -- indeed over one hundred commits..  :(

Yet the good news for you, (and for me), is that once the PR is finally ready and approved by the originating user, the resulting functionality will have been tested far more stringently than it ever had been tested before. :)

**Comment by andrewfg** [comment]
Latest JAR is here...

[org.openhab.binding.velux-3.4.0-SNAPSHOT.jar.zip](https://github.com/openhab/openhab-addons/files/9162837/org.openhab.binding.velux-3.4.0-SNAPSHOT.jar.zip)

**Comment by andrewfg** [comment]
@jlaur for info, the user who requested the enhancement has confirmed the functionality.

File: bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/bridge/slip/FunctionalParameters.java [hunk 16]
```diff
@@ -0,0 +1,202 @@
...
+                requestData.setTwoByteValue(pos, values[i]);
+            }
+            pos = pos + 2;
+            bitMask = bitMask >>> 1;
+        }
+        return fpIndex;
+    }
+
+    @Override
+    public boolean equals(@Nullable Object obj) {
```
Comment (jlaur) [inline bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/bridge/slip/FunctionalParameters.java:190]: When overriding `equals`, you must also override `hashCode`.

File: bundles/org.openhab.binding.velux/src/test/java/org/openhab/binding/velux/test/TestNotificationsAndDatabase.java [hunk 17]
```diff
@@ -0,0 +1,907 @@
...
+
+    private VeluxExistingProducts getExistingProducts() {
+        return EXISTING_PRODUCTS;
+    }
+
+    /**
+     * Confirm the existing products database is initialised.
+     */
+    @Test
+    @Order(1)
```
Comment (jlaur) [inline bundles/org.openhab.binding.velux/src/test/java/org/openhab/binding/velux/test/TestNotificationsAndDatabase.java:85]: Curious, is the order important?

**Review by jlaur — CHANGES_REQUESTED** [review]
Thanks! One comment and one question.

File: bundles/org.openhab.binding.velux/src/test/java/org/openhab/binding/velux/test/TestNotificationsAndDatabase.java (hunk 17 above)
Comment (andrewfg) [reply bundles/org.openhab.binding.velux/src/test/java/org/openhab/binding/velux/test/TestNotificationsAndDatabase.java:85]: Yes. Each test leaves the test article in a given state, that is a prerequisite for the subsequent test.

File: bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/bridge/slip/FunctionalParameters.java (hunk 16 above)
Comment (andrewfg) [reply bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/bridge/slip/FunctionalParameters.java:190]: Ok. Will do.

Comment (andrewfg) [reply bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/bridge/slip/FunctionalParameters.java:190]: Done

**Review by jlaur — APPROVED** [review]
Thanks!

**Comment by jlaur** [comment]
@andrewfg - can I persuade you into preparing a desired commit message for the squash merge? I assume you don't need all 114 commits messages concatenated, so it would be helpful with your own proposal to not mess something up.

**Comment by andrewfg** [comment]
> desired commit message for the squash merge

I suggest the same as the PR name 'Add support for vane/slat position'
